### PR TITLE
feat(YouTube - Save to Watch later): Add support for Shorts

### LINKED ISSUE #2776: feat(YouTube): Add Watch Later option to Shorts three-dot menu in feed
### Feature description

Add a "Watch Later" option to the three-dot menu on Shorts when they show up in the main feed. 
This already exists for long form videos in the feed, and it also already exists for Shorts inside notifications. This request is just to bring that same option to Shorts in the feed too.

### Motivation

**Why should this feature be implemented?**
Right now if a Short shows up in the feed and looks interesting, the only way to save it for later is to open the video and manually add it to a playlist. That is a lot of extra steps for something that takes one tap on long form videos.

**What is the explicit use case?**
Shorts often show up in the feed when I do not have time to watch them right away. I want to be able to tap the three-dot menu on the Short directly in the feed and add it to Watch Later, the same way I already can with regular videos.

**What are the benefits?**
Saves time and cuts out the extra steps of opening the video just to save it. Makes Shorts behave the same way in the feed as they already do everywhere else.

**What makes this feature important?**
The feature is not new to the app, it already works for Shorts in notifications. This is just extending something that already exists to a place where it is currently missing.

### Acknowledgements

- [x] I have checked all open and closed feature requests and this is not a duplicate
- [x] This is a feature request for YouTube, YT Music, or Reddit

### PR DESCRIPTION
**Pull request by sahajsaboo**
Add Watch Later option to Shorts in feed and in Shorts player

### Description

- Adds Watch Later to Shorts feed flyout menus.

- Adds Watch Later to the Shorts player overflow menu.

- Keeps the existing player Watch Later overlay button independent.

- Keeps the feed/player settings independent.

- Extracts shared flyout infrastructure into FlyoutPatch.kt

Closes #2776 

### Additional context

WhatsApp Image 2026-09-13 at 3 30 15 PM
WhatsApp Image 2026-09-13 at 3 30 41 PM
WhatsApp Image 2026-09-13 at 3 30 57 PM
WhatsApp Image 2026-09-13 at 3 31 10 PM

### Test results

- [x] Tested on both the **minimum** and **maximum** supported versions
- [x] Tested on **experimental** supported versions (Optional)

### CODE REVIEW

**Comment by MarcaDian** [comment]
`patches-list.json` automatically generated by a script

**Comment by 0xrxL** [comment]
Be careful with FlyoutUtils class. Every minimal change could break it, in absence of a proper (extended) testing.

**Comment by 0xrxL** [comment]
@tanvir8282 @sr-shishir @realAmanShah @NhaKhanhTran6hz

From the moment that the logic behind the flyout is complex, and you're the best testing team for flyout stuff, maybe you could help this contributor to detect the possible issues. 👍

**Comment by realAmanShah** [comment]
> @tanvir8282 @sr-shishir @realAmanShah @NhaKhanhTran6hz
> 
> From the moment that the logic behind the flyout is complex, and you're the best testing team for flyout stuff, maybe you could help this contributor to detect the possible issues. 👍

Would love to test it, but the GitHub Actions workflow is currently awaiting approval from a maintainer.

<details>
<summary>Screenshots</summary>

Screenshot_20260914_063014_Samsung Browser

</details>

**Comment by 0xrxL** [comment]
> > @tanvir8282 @sr-shishir @realAmanShah @NhaKhanhTran6hz
> > From the moment that the logic behind the flyout is complex, and you're the best testing team for flyout stuff, maybe you could help this contributor to detect the possible issues. 👍
> 
> Would love to test it, but the GitHub Actions workflow is currently awaiting approval from a maintainer.
> 
> Screenshots

Workflow approved.

**Comment by realAmanShah** [comment]
@sahajsaboo Tested the build from the PR artifact. While "Save to Watch later" functions properly, the hook is currently too broad and injects the custom actions into multiple unrelated bottom sheets/menus:

1. **Shorts Remix/Collab menu:** Opens with the Morphe flyout items at the top.
2. **Comment 3-dots popup menu:** Injects the actions directly above "Share / Report".
3. **Shorts Share sheet:** Renders the actions right above the app share sheet.

<details>
<summary>Screenshots</summary>

Screenshot_20260914_072825_YouTube Morphe
Screenshot_20260914_072932_YouTube Morphe
Screenshot_20260914_073001_YouTube Morphe

</details>

**Comment by sahajsaboo** [comment]
I tried a few approaches and spent some time debugging this, but I couldn't find a reliable way to distinguish the Shorts 3-dot menu from the other bottom sheets/menus

**Comment by realAmanShah** [comment]
> I tried a few approaches and spent some time debugging this, but I couldn't find a reliable way to distinguish the Shorts 3-dot menu from the other bottom sheets/menus

Maybe @0xrxL can help with this.

**Comment by 0xrxL** [comment]
> I tried a few approaches and spent some time debugging this, but I couldn't find a reliable way to distinguish the Shorts 3-dot menu from the other bottom sheets/menus

Just don't do it. Check if the current player is shorts or video (then will be obvious that every flyout opened from there, will be part of shorts player). This kind of check is already available in patches.

**Comment by sahajsaboo** [comment]
Got it, I was overcomplicating this way too much. Fixed the issues and pushed the changes. Thanks!

**Comment by 0xrxL** [comment]
> > I tried a few approaches and spent some time debugging this, but I couldn't find a reliable way to distinguish the Shorts 3-dot menu from the other bottom sheets/menus
> 
> Maybe @0xrxL can help with this.

Try the last artifact. 👍

File: extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/utils/FlyoutUtils.java [hunk 1]
```diff
@@ -96,6 +96,25 @@ public record FlyoutMenuInfo(
...
+            getAsciiBytes("shorts_video_cell.e"),
+            getAsciiBytes("swipeable_row.e"),
+            getAsciiBytes("video_lockup_with_attachment.e")
+    );
+
+    private static final List<byte[]> SHORTS_VIDEO_ELEMENTS_BYTES = List.of(
+            getAsciiBytes("shorts_pivot_item.e"),
+            getAsciiBytes("shorts_shelf.e"),
+            getAsciiBytes("shorts_video_cell.e")
+    );
```
[the remark this answers is not in the document]
Comment (0xrxL) [reply extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/utils/FlyoutUtils.java:107]: @sahajsaboo These filters are not needed at all. The various extractors (videoId, commentId, playlistId etc.) already have the needed checks to fetch the related Ids without throw any exception.

Leave only the check about the shorts player. 👍

**Comment by realAmanShah** [comment]
Tested last artifact:

**What's fixed:**
- Unwanted flyout injections in the Remix sheet, Comment menu, and Share sheet are resolved.
- "Save to Watch later" functions as expected.

**New regressions introduced in this build:**

1. **Whitelist items missing on cold start:**
   - After a force stop/clearing cache, tapping the 3 dots on a Short in the feed for the first time does not show the whitelist actions at all (only "Save to Watch later" and default items appear).
   - Closing and reopening the same flyout menu makes them appear.

<details>
<summary>Screenshots</summary>

Screenshot_20260915_072521

</details>


2. **Whitelist state desync across feed items:**
   - In the feed Shorts shelf, if you tap 3 dots on one Short and add it to the whitelist, tapping the 3 dots on adjacent or other Shorts in the shelf incorrectly shows "Remove channel from whitelist" instead of "Add", even though those channels were never whitelisted.
   - Playing the Short and opening the in-player flyout displays the correct state.

<details>
<summary>Screenshots</summary>

Screenshot_20260915_071525
Screenshot_20260915_071555
Screenshot_20260915_071619


</details>

**Comment by NhaKhanhTran6hz** [comment]
also, this flyout should not appear on this, btw the flyout should only show flyout on single watched short, not multiple.
<details>
Screenshot_2026-09-15-16-19-49-30

**Comment by sahajsaboo** [comment]
@NhaKhanhTran6hz that flyout is present in regular (1.42.0-dev.3) version of morphe too 

<details>
WhatsApp Image 2026-09-15 at 3 20 24 PM(1)
WhatsApp Image 2026-09-15 at 3 20 24 PM

**Comment by sahajsaboo** [comment]
in this latest push i have -

- Fixed the Whitelist items missing on cold start and the Whitelist state desync across feed items by removing the Whitelist options from the Shorts feed 3-dot menu, matching stable Morphe where these options already don't exist. This also seems unnecessary for Shorts since Shorts revenue works differently (Shorts Fund).
- Kept the Whitelist options in the Shorts player as they are in stable Morphe.
- Added Queue to Shorts in the feed, matching stable Morphe behavior.
- Removed extra/unnecessary filters.
- Fixed the other issues mentioned above (apart from the existing behavior of the 3-dot flyout appearing for multiple Shorts in History since this is not specific to this PR and already exists in current Morphe versions)

**Comment by 0xrxL** [comment]
@sahajsaboo Remember: if you'll break FlyoutUtils, I'll not fix it again. 😂

**Comment by 0xrxL** [comment]
@realAmanShah Oh sorry. I approved the workflow, you can now try it. 👍

**Comment by realAmanShah** [comment]
> @realAmanShah Oh sorry. I approved the workflow, you can now try it. 👍

@0xrxL Tested the latest artifact and everything in this PR is working cleanly now.

Just a heads-up regarding that other behavior: in the **You** tab under the **History** carousel, tapping the 3 dots next to the grouped "Shorts" header still injects the channel whitelist options. Since it's a multi-video container and not a single Short, showing channel whitelist actions there is ambiguous. As @sahajsaboo pointed out, this is pre-existing in stable Morphe and unrelated to this PR, so this PR looks good to merge!

<details>
<summary>Screenshots</summary>

Screenshot_20260916_065541_YouTube Morphe

</details>

**Comment by 0xrxL** [comment]
> > @realAmanShah Oh sorry. I approved the workflow, you can now try it. 👍
> 
> @0xrxL Tested the latest artifact and everything in this PR is working cleanly now.
> 
> Just a heads-up regarding that other behavior: in the **You** tab under the **History** carousel, tapping the 3 dots next to the grouped "Shorts" header still injects the channel whitelist options. Since it's a multi-video container and not a single Short, showing channel whitelist actions there is ambiguous. As @sahajsaboo pointed out, this is pre-existing in stable Morphe and unrelated to this PR, so this PR looks good to merge!
> 
> Screenshots

I'll push a commit here asap, to fix this issue. 👍

**Comment by 0xrxL** [comment]
@realAmanShah Should be fixed now. 👍

**Comment by realAmanShah** [comment]
> @realAmanShah Should be fixed now. 👍

@0xrxL Tested last artifact, but the regressions are still present along with a missing action:

1. **"Save to Watch later" missing from feed shelf:**
   - The 3-dot menu on Shorts in the Home feed no longer shows "Save to Watch later" at all (it only appears inside the Shorts player).

<details>
<summary>Screenshots</summary>

Screenshot_20260916_133849_YouTube Morphe

</details>

2. **Whitelist state desync across feed items still occurs:**
   - In the feed Shorts shelf, if you tap 3 dots on one Short and add it to the whitelist, tapping the 3 dots on adjacent or other Shorts in the shelf incorrectly shows "Remove channel from whitelist" instead of "Add", even though those channels were never whitelisted.

<details>
<summary>Screenshots</summary>

Screenshot_20260916_134219_YouTube Morphe
Screenshot_20260916_134229_YouTube Morphe

</details>

3. **History carousel container still has whitelist options:**
   - In `You > History`, tapping the 3 dots next to the grouped "Shorts" header still injects the channel whitelist actions into that shelf container menu.

<details>
<summary>Screenshots</summary>

Screenshot_20260916_134305_YouTube Morphe

</details>

**Comment by dr-strange15** [comment]
unrelated but whos gonna save a 30-60 seconds short to watch it later,lol😅
maybe a usefull short

**Comment by 0xrxL** [comment]
@realAmanShah Try again :+1:

**Comment by 0xrxL** [comment]
@sahajsaboo Please, do not touch the code furtherly. As you can see, we're reviewing it, and will fix any possible issue on our side.

Only a remind. 👍

Regards.

**Comment by realAmanShah** [comment]
@0xrxL Tested last artifact:

### Working as intended:
- **Shorts shelf (Feed):** 3-dot menu opens cleanly, no whitelist desync between items, and actions function properly.
- **Shorts player (only with 3-dot menu):** All custom actions (Whitelist and "Save to Watch later") appear and work correctly.

### Observations/Edge Cases:
1. **Shorts player (Long-press flyout):** 
   - Long-pressing on the player opens the stock bottom sheet (Description, Quality, etc.) without any custom Morphe injections. Only the top-right 3-dot button gets injected.

<details>
<summary>Screenshots</summary>

Screenshot_20260917_065901
Screenshot_20260917_070035

</details>

2. **Standard video player (Gear menu):** 
   - "Save to Watch later" now also appears inside the player gear settings menu alongside Quality/Playback speed.

<details>
<summary>Screenshots</summary>

Screenshot_20260917_065324

</details>

3. **History carousel (You tab):**
    - Tapping the 3 dots on Shorts/videos does not open any Morphe custom flyout actions at all, cleanly bypassing the previous ambiguous container injection.

<details>
<summary>Screenshots</summary>

Screenshot_20260917_070530

</details>


Overall, the primary regressions and state desyncs are resolved! 👍

**Comment by 0xrxL** [comment]
> Overall, the primary regressions and state desyncs are resolved! 👍

Added another setting to fix the second edge case.

I'll check the first one, to close the situation.

Thanks for your testing efforts. 👍

**Comment by 0xrxL** [comment]
@realAmanShah First edge case fixed too.

Make a last test before merge. 👍

**Comment by realAmanShah** [comment]
> @realAmanShah First edge case fixed too.
> 
> Make a last test before merge. 👍

@0xrxL Tested last artifact:

- **Shorts shelf (Feed):** Working cleanly. "Save to Watch later" appears and functions properly with no whitelist desync.
- **Shorts player (Long-press & 3-dot):** Whitelist actions appear on both menus now.
- **Watch Later in player:** "Save to Watch later" is indeed not in the Shorts player flyout, but honestly **there is no need for it there**. The player already has YouTube's native on-screen "Save" button (and long-pressing it opens "Save to Watch later" anyway). Having it in the feed shelf was the missing piece.

Everything looks solid and clean. Ready to merge! 🚀

**Comment by 0xrxL** [comment]
@sahajsaboo Thanks for your contribution. ❤️
